File: 13_1-D_Dynamic_Programming/01_Climbing_Stairs/0070-climbing-stairs.cpp
/*
Problem: LeetCode 70 - Climbing Stairs

Description:
You are climbing a staircase that has n steps. You can either climb 1 or 2 steps at a time.
Return the number of distinct ways to climb to the top.

Intuition:
To reach the nth step, we can either take a single step from the (n-1)th step or take two steps from the (n-2)th step.
This forms the basis of our dynamic programming approach, as we can break down the problem into subproblems and build the solution from there.

Approach:
1. Initialize an array dp of size (n+1) to store the number of distinct ways to reach each step.
2. Set the base cases: dp[0] = 1 (no steps needed) and dp[1] = 1 (one step to reach the first step).
3. Iterate from 2 to n:
   - Compute dp[i] by summing up the number of ways to reach the previous two steps: dp[i] = dp[i-1] + dp[i-2].
4. Return dp[n], which represents the number of distinct ways to reach the top step.

Time Complexity:
The time complexity is O(n) since we iterate through the steps once.

Space Complexity:
The space complexity is O(n) since we use an array of size (n+1) to store the intermediate results.

Dynamic Programming:
- Subproblem: The subproblem is finding the number of distinct ways to reach the current step.
- Recurrence Relation: dp[i] = dp[i-1] + dp[i-2], where dp[i] represents the number of distinct ways to reach the ith step.
- Base Case: dp[0] = 1 and dp[1] = 1, as there is only one way to reach the first two steps.
*/

class Solution {
  public:
    int climbStairs(int n) {
        if (n <= 2) {
            return n;  // Base cases
        }

        vector<int> dp(n + 1);
        dp[0] = 1;
        dp[1] = 1;

        for (int i = 2; i <= n; ++i) {
            dp[i] = dp[i - 1] + dp[i - 2];  // Recurrence relation
        }

        return dp[n];
    }
};

// O(1) Space
// class Solution {
// public:
//     int climbStairs(int n) {
//         if(n <= 2)
//             return n;

//         int first = 1, second = 2;

//         for(int  i = 2; i < n; i++) {
//             int temp = second;
//             second = first + second;
//             first = temp;
//         }
//         return second;
//     }
// };

File: 13_1-D_Dynamic_Programming/01_Climbing_Stairs/0070-climbing-stairs.rs
/*
Problem: LeetCode 70 - Climbing Stairs

Key Idea:
The key idea is to realize that to reach the i-th step, you can either:
1. Take one step from the (i-1)-th step if you had already climbed (i-1) steps.
2. Take two steps from the (i-2)-th step if you had already climbed (i-2) steps.

Approach:
1. Initialize two variables, `prev` and `curr`, to represent the number of ways to reach the (i-1)-th and i-th step, respectively.
2. Start iterating from step 3 to n (inclusive), where n is the total number of steps:
   - Calculate the number of ways to reach the i-th step as `curr = prev + curr`, where `prev` represents the number of ways to reach the (i-1)-th step, and `curr` represents the number of ways to reach the i-th step.
   - Update `prev` to the previous value of `curr`.
3. Return the value of `curr` as the number of ways to reach the n-th step.

Time Complexity:
O(n), where n is the total number of steps. We perform a single pass through the steps.

Space Complexity:
O(1), as we use only two variables (`prev` and `curr`) to store intermediate results.
*/

impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        if n <= 2 {
            return n;
        }

        let (mut prev, mut curr) = (1, 2);

        for _ in 3..=n {
            let temp = curr;
            curr = prev + curr;
            prev = temp;
        }

        curr
    }
}
